org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractQuickfixTest.xtend

```diff
@@ -0,0 +1,194 @@
+import org.eclipse.xtext.ui.editor.XtextEditor
+import org.eclipse.xtext.ui.editor.XtextEditorInfo
+import org.eclipse.xtext.ui.editor.model.IXtextDocument
+import org.eclipse.xtext.ui.editor.model.XtextDocument
+import org.eclipse.xtext.ui.editor.model.edit.IModificationContext
+import org.eclipse.xtext.ui.editor.quickfix.IssueResolution
+import org.eclipse.xtext.ui.editor.quickfix.IssueResolutionProvider
+import org.eclipse.xtext.ui.refactoring.ui.SyncUtil
+import org.eclipse.xtext.ui.testing.util.IResourcesSetupUtil
+import org.eclipse.xtext.util.CancelIndicator
+import org.eclipse.xtext.util.StringInputStream
+import org.eclipse.xtext.util.Strings
+import org.eclipse.xtext.validation.CheckMode
+import org.eclipse.xtext.validation.IResourceValidator
+
+import static extension org.eclipse.xtext.ui.testing.util.IResourcesSetupUtil.addNature
+
+/**
+ * @since 2.16
+ */
+abstract class AbstractQuickfixTest extends AbstractEditorTest {
+
+	@Inject protected Injector injector
+	@Inject protected XtextEditorInfo editorInfo
+
+	@Inject protected extension SyncUtil
+	@Inject protected extension IResourceValidator
+	@Inject protected extension FileExtensionProvider
+	@Inject protected extension IssueResolutionProvider
+
+	override protected getEditorId() {
+		editorInfo.getEditorId
+	}
+
+	/**
+	  * Test that the expected quickfixes are offered on a given validation issue in a given dsl text.
+	  * 
+	  * @param it The initial dsl text.
+	  * @param issueCode The code of the validation issue to that the offered quickfixes are to test.
+	  * @param quickfixes The quickfixes that are expected to be offered on the given <code>issueCode</code>.
+	  * Each expected quickfix should be described by the following triple:
+	  *  1. the quickfix label
+	  *  2. the quickfix description
+	  *  3. the dsl text after the quickfix application
+	  */
+	def testQuickfixesOn(CharSequence it, String issueCode, List<List<String>> quickfixes) {
+		// given
+		dslFile.
+		// when
+		openInEditor.
+		// then
+		quickfixesAreOffered(issueCode, quickfixes)
+	}
+
+	protected def dslFile(CharSequence content) {
+		val file = IResourcesSetupUtil.createFile(projectName, fileName, fileExtension, content.toString)
+		
+		/*
+		 * TODO: find a better (with good performance) solution
+		 * to set the Xtext nature on the test project.
+		 */
+		val project = file.project
+		if(!project.hasNature(XtextProjectHelper.NATURE_ID)) {
+			project.addNature(XtextProjectHelper.NATURE_ID)
+		}
+		
+		file
+	}
+
+	protected def String getProjectName() '''QuickfixTestProject'''
+
+	protected def String getFileName() '''quickfix'''
+
+	protected def getFileExtension() {
+		primaryFileExtension
+	}
+
+	protected def openInEditor(IFile dslFile) {
+		/*
+		 * wait for the cross-reference resolution
+		 */
+		new NullProgressMonitor().waitForBuild
+		dslFile.openEditor
+	}
+
+	protected def quickfixesAreOffered(XtextEditor editor, String issueCode, List<List<String>> expected) {
```

Comment:
I think List<List<String>> is complex to understand for the user how to use this. Maybe you find a smoother solution, e.g. with a dedicated data structure?